### DISCUSSION TITLE
How to source quarkus properties from a custom ConfigSource?

### DISCUSSION THREAD

**Discussion by rsshah** [quarkusio/quarkus#31915] [Q&A] [answered]
I have a custom `ConfigSource` which doesn't actually get properties until after runtime. The initialization flow looks like:

1. Load initial runtime properties passed in through the default ConfigSources (file, env variable, system, etc)
2. Create the custom ConfigSource
3. Create a client to a config service
4. fetch properties from the service and load them into the custom ConfigSource

I want some of those properties to be quarkus extension configuration properties - for example, things like an OIDC url and credentials to use in the oidc-client.

Is this kind of dynamic property pattern supported in Quarkus?

Right now my custom ConfigSource does populate the property system with properties fetched from the service, but I want to delay the initialization of extensions until all properties have been loaded by all ConfigSources.

**Comment by radcortez** [comment]
You can use a `ConfigSourceFactory`. Please check: https://smallrye.io/smallrye-config/3.2.0/config-sources/factories/.

**Reply by arinn1204** [reply]
hey @radcortez, I have a similar use case and I want to be able to inject CDI beans into my custom `ConfigSource`/`ConfigSourceFactory`. Otherwise, I would need to create duplicates of these beans at runtime in SPI, which feels like a hack. What's the best way to do this?

My custom ConfigSource would use the beans to create clients and make requests against external services to source the properties.

**Reply by radcortez** [reply, marked as answer]
I do not recommend mixing CDI Beans with ConfigSources. The `Config` is one of the first things we create in Quarkus to configure the runtime (including CDI). If we allow CDI Injection in sources, you get a chicken and egg problem. You need CDI to be ready to inject the bean, but you need Config to be ready to configure CDI.

If you are absolutely sure that your source will only get called when CDI is already up (well, the source had to be created and available in `Config`), you could programmatically retrieve CDI and use it, but again, this is not recommended.
